File: case_studies/wtp/runner_utils_test.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
import os
import unittest
from unittest.mock import patch

from case_studies.wtp.models import custom_types
from case_studies.wtp import runner_utils


class RunnerUtilsTest(unittest.TestCase):

  def run_generate_and_save_topic_tree(self, topic_tree_data) -> str:
    """Wrapper for generate_and_save_topic_tree that reads output file and returns str."""

    output_file_base = "test_topic_tree"

    with patch("builtins.print"):  # suppress output prints
      runner_utils.generate_and_save_topic_tree(
          topic_tree_data, output_file_base
      )

    # Check if the TXT file was created and has the correct content
    txt_file_path = f"{output_file_base}.txt"
    self.assertTrue(os.path.exists(txt_file_path))

    with open(txt_file_path, "r", encoding="utf-8") as f:
      content = f.read()

    # Clean up the created files
    os.remove(txt_file_path)

    return content

  def test_generate_and_save_topic_tree(self):
    tree_str = self.run_generate_and_save_topic_tree([{
        "topic_name": "Topic 1",
        "opinions": [
            {
                "opinion_text": "Opinion 1.1",
                "representative_texts": ["Quote 1.1"],
            },
            {
                "opinion_text": "Opinion 1.2",
                "representative_texts": ["Quote 1.2"],
            },
        ],
    }])
    self.assertEqual(tree_str, (
        "1. Topic 1 (2 quotes)\n"
        "  1. Opinion 1.1 (1 quotes)\n"
        "  2. Opinion 1.2 (1 quotes)\n\n"
        "Total number of unique opinions: 2"
    ))

  def test_generate_and_save_topic_tree_with_duplicate_quotes(self):
    tree_str = self.run_generate_and_save_topic_tree([{
        "topic_name": "Topic 1",
        "opinions": [
            {
                "opinion_text": "Opinion 1.1",
                "representative_texts": ["Quote 1.1"],
            },
            {
                "opinion_text": "Opinion 1.2",
                "representative_texts": ["Quote 1.1", "Quote 1.2"],
            },
        ],
    }])
    self.assertEqual(tree_str, (
        "1. Topic 1 (2 quotes)\n"
        "  1. Opinion 1.2 (2 quotes)\n"
        "  2. Opinion 1.1 (1 quotes)\n\n"
        "Total number of unique opinions: 2"
    ))


if __name__ == "__main__":
  unittest.main()

File: case_studies/wcbgb/library/runner-cli/runner_utils.test.ts
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { getCommentsFromCsv, parseTopicsString } from "./runner_utils";
import { Comment, VoteTally } from "../src/types";
import { Readable } from "stream";

// Mock FileStream to be able to test the reading of CSVs.
jest.mock("fs", () => {
  const actualFs = jest.requireActual("fs");
  let mockCsvData = "";
  let mockHeaderData = "";

  const mockCreateReadStream = jest.fn().mockImplementation(() => {
    const stream = new Readable();
    stream.push(mockCsvData);
    stream.push(null);
    return stream;
  });
  const mockReadFileSync = jest.fn().mockImplementation(() => {
    return mockHeaderData;
  });

  return {
    ...actualFs,
    readFileSync: mockReadFileSync,
    createReadStream: mockCreateReadStream,
    __setMockCsvData: (data: string) => {
      mockCsvData = data;
    },
    __setMockHeaderData: (data: string) => {
      mockHeaderData = data;
    },
  };
});

describe("getCommentsFromCsv", () => {
  const mockFilePath = "mock-file.csv";
  let mockFs: { [key: string]: jest.Mock };

  beforeEach(() => {
    mockFs = jest.requireMock("fs");
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should read comments with group vote tallies from a CSV file", async () => {
    const mockCsvContent = `comment-id,comment_text,group-0-agree-count,group-0-disagree-count,group-0-pass-count,group-1-agree-count,group-1-disagree-count,group-1-pass-count
1,comment1,10,5,0,5,10,5
2,comment2,2,5,3,5,3,2`;
    const mockHeader = mockCsvContent.split("\n")[0];
    mockFs.__setMockHeaderData(mockHeader);
    mockFs.__setMockCsvData(mockCsvContent);

    const comments: Comment[] = await getCommentsFromCsv(mockFilePath);

    expect(comments).toHaveLength(2);

    expect(comments[0].id).toBe("1");
    expect(comments[0].text).toBe("comment1");
    expect(comments[0].voteInfo).toEqual({
      "group-0": new VoteTally(10, 5, 0),
      "group-1": new VoteTally(5, 10, 5),
    });

    expect(comments[1].id).toBe("2");
    expect(comments[1].text).toBe("comment2");
    expect(comments[1].voteInfo).toEqual({
      "group-0": new VoteTally(2, 5, 3),
      "group-1": new VoteTally(5, 3, 2),
    });
  });

  it("should read comments with no group vote tallies from a CSV file", async () => {
    const mockCsvContent = `comment-id,comment_text,agrees,disagrees,passes
1,comment1,10,5,0
2,comment2,2,5,3`;
    mockFs.__setMockHeaderData(mockCsvContent.split("\n")[0]);
    mockFs.__setMockCsvData(mockCsvContent);

    const comments: Comment[] = await getCommentsFromCsv(mockFilePath);

    expect(comments).toHaveLength(2);

    expect(comments[0].id).toBe("1");
    expect(comments[0].text).toBe("comment1");
    expect(comments[0].voteInfo).toEqual(new VoteTally(10, 5, 0));

    expect(comments[1].id).toBe("2");
    expect(comments[1].text).toBe("comment2");
    expect(comments[1].voteInfo).toEqual(new VoteTally(2, 5, 3));
  });
});

describe("parseTopicsString", () => {
  it("should parse a single topic string", () => {
    const topicsString = "Topic A:Subtopic A.1";
    const expectedTopics = [
      { name: "Topic A", subtopics: [{ name: "Subtopic A.1", subtopics: [] }] },
    ];
    expect(parseTopicsString(topicsString)).toEqual(expectedTopics);
  });

  it("should parse multiple topic strings", () => {
    const topicsString = "Topic A:Subtopic A.1;Topic B:Subtopic B.1;Topic C";
    const expectedTopics = [
      { name: "Topic A", subtopics: [{ name: "Subtopic A.1", subtopics: [] }] },
      { name: "Topic B", subtopics: [{ name: "Subtopic B.1", subtopics: [] }] },
      { name: "Topic C" },
    ];
    expect(parseTopicsString(topicsString)).toEqual(expectedTopics);
  });

  it("should handle topic strings with only topic names", () => {
    const topicsString = "Topic A;Topic B;Topic C";
    const expectedTopics = [{ name: "Topic A" }, { name: "Topic B" }, { name: "Topic C" }];
    expect(parseTopicsString(topicsString)).toEqual(expectedTopics);
  });

  it("should handle topic strings with only topic names, including : separators", () => {
    const topicsString = "Topic A:;Topic B:;Topic C:";
    const expectedTopics = [{ name: "Topic A" }, { name: "Topic B" }, { name: "Topic C" }];
    expect(parseTopicsString(topicsString)).toEqual(expectedTopics);
  });

  it("should handle topic strings with only subtopic names", () => {
    const topicsString = "Topic A:Subtopic A.1;Topic B:Subtopic B.1";
    const expectedTopics = [
      { name: "Topic A", subtopics: [{ name: "Subtopic A.1", subtopics: [] }] },
      { name: "Topic B", subtopics: [{ name: "Subtopic B.1", subtopics: [] }] },
    ];
    expect(parseTopicsString(topicsString)).toEqual(expectedTopics);
  });

  it("should handle topic strings with multiple subtopics", () => {
    const topicsString =
      "Topic A:Subtopic A.1;Topic A:Subtopic A.2;Topic B:Subtopic B.1;Topic B:Subtopic B.2";
    const expectedTopics = [
      {
        name: "Topic A",
        subtopics: [
          { name: "Subtopic A.1", subtopics: [] },
          { name: "Subtopic A.2", subtopics: [] },
        ],
      },
      {
        name: "Topic B",
        subtopics: [
          { name: "Subtopic B.1", subtopics: [] },
          { name: "Subtopic B.2", subtopics: [] },
        ],
      },
    ];
    expect(parseTopicsString(topicsString)).toEqual(expectedTopics);
  });

  it("should handle topic strings with themes", () => {
    const topicsString = "Topic A:Subtopic A:Theme A;Topic A:Subtopic A:Theme B";
    const expectedTopics = [
      {
        name: "Topic A",
        subtopics: [{ name: "Subtopic A", subtopics: [{ name: "Theme A" }, { name: "Theme B" }] }],
      },
    ];

    expect(parseTopicsString(topicsString)).toEqual(expectedTopics);
  });
});
